Get themes from all GTK theme directories

Winter-Knight opened the pull request.

Winter-Knight committed:
* add themes in /usr/local/share/themes into system theme list

Winter-Knight committed:
* use theme directory for gtk2 widget factory
* use all theme directories for gtk3 widget factory and remove duplicates

Winter-Knight committed:
* awf-gtk2 loads themes from gtk_rc_get_theme_dir()
* awf-gtk3 loads themes from all theme directories
* awf-gtk3 uses hashtable as a set to ensure unique entries and for quicker searching
* awf-gtk4 loads themes from a few directories, more work needed
* fixed failure to compile when optimization is disabled by adding locale.h to each source file

luigifab commented:
With the attached source files, from an updated version of your changes:

```
fabrice@debian:~/gtk-awf$ ./awf-gtk2
/usr/share/mate/themes
/usr/local/share/themes
/usr/share/themes
/home/fabrice/.local/share/themes
/home/fabrice/.themes
fabrice@debian:~/gtk-awf$ ./awf-gtk3
/usr/share/mate/themes
/usr/local/share/themes
/usr/share/themes
/home/fabrice/.local/share/themes
/home/fabrice/.themes
fabrice@debian:~/gtk-awf$ ./awf-gtk4
/usr/share/mate/themes
/usr/local/share/themes
/usr/share/themes
/home/fabrice/.local/share/themes
/home/fabrice/.themes
```

It seems that it works! Is it working for you?
<del>The last? thing to fix, is [g_hash_table_add](https://docs.gtk.org/glib/struct.HashTable.html) available only from 2.32, and but I must remain compatible from 2.28.</del> -> ChatGPT found a way.

---

[awf-gtk2.c](https://github.com/user-attachments/files/22197248/awf-gtk2.c)
[awf-gtk3.c](https://github.com/user-attachments/files/22197249/awf-gtk3.c)
[awf-gtk4.c](https://github.com/user-attachments/files/22197250/awf-gtk4.c)

Winter-Knight committed:
* Incorporated Luigifab's changes from here: https://github.com/luigifab/awf-extended/pull/10#issuecomment-3263830717

Winter-Knight committed:
* awf-gtk2 includes only theme directories that gtk2 searches
* awf-gtk3 and awf-gtk4 use hash table for faster searching

Winter-Knight commented:
Awesome!

A few notes:

locale - I noticed you removed #include <locale.h>. I know it's not really relevant to this patch but it is important. You call setlocale (LC_ALL, "") and that function and enum are both declared in locale.h. Your project compiles anyway, but when optimizations are turned off it does not.

gtk2 doesn't have the same theme dirs as gtk3 and gtk4. You include files from local system data directories but gtk2 does not, so if you try to load those themes in awf-gtk2 then the active theme just appears to be Raleigh.

I recommend the changes in my newly updated useallthemedirs branch. I've reverted awf-gtk2 to only load the themes that will work, and I've updated each of the source files to use the hash table for searching and they are compatible with glib 2.28.

luigifab commented:
Ok understood.
You are right for gtk2.

I see you have replaced all `g_slist_find_custom` by `g_hash_table_lookup`, this add a new global variable, maybe it's better?

Winter-Knight commented:
Yes, I think it's better. Searching through a hash tree is faster than
searching through a linked list.

luigifab committed:
Remove unrelated changes and readd locale.h

luigifab committed:
Remove unrelated changes and readd locale.h

luigifab committed:
Remove unrelated changes and readd locale.h

luigifab reviewed (approved):
Working, will be merged and added for the next release next month.

Winter-Knight commented:
Awesome! Thanks.
